pkg/catalogv2/oci/oci.go

```diff
@@ -0,0 +1,233 @@
+package oci
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/Masterminds/semver"
+	ocispecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	"github.com/sirupsen/logrus"
+	"helm.sh/helm/v3/pkg/registry"
+	"helm.sh/helm/v3/pkg/repo"
+	corev1 "k8s.io/api/core/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/content/memory"
+	"oras.land/oras-go/v2/registry/remote"
+
+	v1 "github.com/rancher/rancher/pkg/apis/catalog.cattle.io/v1"
+)
+
+// Chart returns an io.ReadCloser of the chart tar that is requested.
+// It uses oras Go library to download the manifest of the OCI artifiact
+// checks if it is a Helm chart and then return the chart tar layer.
+func Chart(credentialSecret *corev1.Secret, chart *repo.ChartVersion, URL string, clusterRepoSpec v1.RepoSpec) (io.ReadCloser, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	if len(chart.URLs) == 0 {
+		return nil, fmt.Errorf("failed to find chartName %s version %s: %w", chart.Name, chart.Version, validation.NotFound)
```

Comment:
I think the error here should mention that every chart should have at least one URL

Reply:
Done.